include/lsst/afw/image/TransmissionCurve.h

```diff
@@ -0,0 +1,300 @@
+     */
+    static std::shared_ptr<TransmissionCurve> makeIdentity();
+
+    /**
+     *  Create a new TransmissionCurve with spatially-constant throughput.
+     *
+     *  @param[in]  throughput      an Array of throughput values with the same
+     *                              size as the wavelengths Array (will be copied).
+     *  @param[in]  wavelengths     an Array of wavelengths in Angstroms (will be copied).
+     *                              Must be monotonically increasing.
+     *  @param[in]  throughputAtMin the throughput value used for wavelengths
+     *                              below wavelengths.front().
+     *  @param[in]  throughputAtMax the throughput value used for wavelengths
+     *                              above wavelengths.back().
+     */
+    static std::shared_ptr<TransmissionCurve> makeConstant(
+        ndarray::Array<double const,1> const & throughput,
+        ndarray::Array<double const,1> const & wavelengths,
+        double throughputAtMin=0.0, double throughputAtMax=0.0
+    );
+
+    /**
+     *  Create a new TransmissionCurve with throughput varying as function of radius.
+     *
+     *  @param[in]  throughput      an Array of throughput values with shape
+     *                              (wavelengths.size(), radii.size()).  Will be
+     *                              copied.
+     *  @param[in]  wavelengths     an Array of wavelengths in Angstroms (will be copied).
+     *                              Must be monotonically increasing.
+     *  @param[in]  radii           an Array of radii (will be copied).
+     *                              Must be monotonically increasing.
+     *  @param[in]  throughputAtMin the throughput value used for wavelengths
+     *                              below wavelengths.front().
+     *  @param[in]  throughputAtMax the throughput value used for wavelengths
+     *                              above wavelengths.back().
+     *
+     *  Throughput at radii below radii.front() or above radii.back() is assumed
+     *  to be constant.
+     */
+    static std::shared_ptr<TransmissionCurve> makeRadial(
+        ndarray::Array<double const,2> const & throughput,
+        ndarray::Array<double const,1> const & wavelengths,
+        ndarray::Array<double const,1> const & radii,
+        double throughputAtMin=0.0, double throughputAtMax=0.0
+    );
+
+    /**
+     *  Return a new TransmissionCurve that simply multiplies the values of two others.
+     *
+     *  The product is typically computed lazily (the returned object samples
+     *  from each operand, and then multiplies those samples).
+     *
+     *  @note This is mapped to __mul__ in Python (overriding operator* in C++
+     *        would be problematic due to the use of shared_ptr).
+     */
+    static std::shared_ptr<TransmissionCurve> multiply(
+        std::shared_ptr<TransmissionCurve> a,
+        std::shared_ptr<TransmissionCurve> b
```

Comment:
I think it's all a package deal - given that I *have* enforced that these objects always be held by `shared_ptr`, `shared_ptr` arguments aren't actually a burden, because anyone holding a `TransmissionCurve` should have a full `shared_ptr`, not just reference.  Take out one link in the chain (immutable+noncopyable+always shared_ptr) and it does all fall apart, but together it does make sense (and is slightly more Python-friendly, IMO).

Reply:
I will say that it drives me nuts to have shared_ptr arguments when references will do. It enables passing an empty pointer, which has to be guarded against, and makes it a hassle to pass in a default-constructed argument when that is appropriate. This is a real nuisance in PropertySet and I just ran into it again today. Personally I would *much* rather just dereference as needed. I realize you disagree about that.

Reply:
For `PropertySet` the argument is very different, because it is not immutable.  There the whole chain of logic falls apart, and I completely agree that pass by [const] reference is appropriate.

Reply:
I'm not going to push more on the use of `shared_ptr` itself, but @r-owen raises a good point about nulls. I did not consider that when looking at the code, so please make sure the behavior is sensible (including test coverage).